### DISCUSSION TITLE
Landlock development discussion

### DISCUSSION THREAD

**Discussion by netblue30** [netblue30/firejail#6065] [General]
I've just cut a branch for landlock. I'll start re-merging pull request #5315. Stay away from the branch for now, things will break! Will go live in the next release (probably January).

Relates to:

* #6078

**Comment by kmk3** [comment]
> I've just cut a branch for landlock. I'll start re-merging pull request #5315
> from @ChrysoliteAzalea. Stay away from the branch for now, things will break!
> Will go live in the next release (probably January).

Thanks for using a separate branch for this!

Please open a PR whenever you think it's ready for review, as there were many
fixes/improvements to be done (at least in the original version).

In the PR, I can do these fixes directly in the branch as well if you want.

**Reply by netblue30** [reply]
Sure, will do! It will be ready to merge in about one week.

**Comment by curiosityseeker** [comment]
This is certainly exciting news! However, it should be noted that the birdcage project recently [removed](https://github.com/phylum-dev/birdcage/commit/18112cbbee15d051d17016144d04f13b4fb94514) Landlock: 

>  Remove landlock

> This patch completely removes landlock from the Linux sandbox.

> While landlock provides solid filesystem isolation even without
> requiring any intricate Linux sandboxing knowledge, it is currently
> still too limited to build a "bulletproof" filesystem sandbox.

> As such, it is better suited for best-effort isolation of "assumed safe"
> applications, rather than sandboxing of "potentially hazardous"
> software.

So ... is including Landlock possibly a bit pre-mature? Or are the existing Landlock features already worth enough to be added to Firejail?

**Reply by rusty-snake** [reply]
>  it is currently still too limited to build a "bulletproof" filesystem sandbox.

Reads as a "landlock alone is to limited" but we combine it with namespaces, seccomp, capabilities, ...

> Or are the existing Landlock features already worth enough to be added to Firejail?

That's the question. Depening on how we add it and how/what we expose to profiles. In the following points Landlock is already superior to mount-namespace features:

- Make `/` noexec and allow exec only for `/usr/{lib,lib64,bin,sbin}` or even only `/usr/{lib,lib64,bin/program,bin/program2}`. (And forbid write on those paths to get a W^X filesystem).
- Allow to read and write to files in a directory but not read (list) a directory itself. i.e. `ls /etc` is forbided, but `cat /etc/resolv.conf` allowed.
- Allow to write to existing files but not to create new ones.

For stuff like making paths readonly it has no benefit beyond "hey we use Landlock" as long as the fundamental architecture does not change.

**Reply by curiosityseeker** [reply]
Thanks for this detailed reply! You should [know](https://codeberg.org/crabjail/crablock) what you're talking about ... :+1:

**Comment by netblue30** [comment]
OK, this is what I have in this moment:
 * Compile time detection based on /usr/include/kernel/landlock.h - if the file is present in the filesystem, the feature is compiled in.
 * Run-time detection of kernels 6.1 (debian stable) or newer - maybe we can push it down to 5.13, but we need to test it.
 * Landlock is the last OS filter applied to the sandbox. All other existing features should work with landlock set on top of them. For example you run "firejail --landlock firefox" and it should work as before.
 * The default filesystem set by landlock is very similar to what we have today in apparmor: full system read, some directories read-write, exec only on well-known exec system directories. We are missing some features in the kernel to make it a full replacement for apparmor, but should do it for now. The filesystem code is here: https://github.com/netblue30/firejail/blob/landlock/src/firejail/landlock.c#L189

I think I'll merge it over the weekend.

**Reply by rusty-snake** [reply]
> Run-time detection of kernels 6.1 

Why 6.1? Landlock had nothing new in the 6.1 release.

> maybe we can push it down to 5.13

We should be able. Landlock is designed with up/downward compatibility and we target ABI version 1 ATM.

> Landlock is the last OS filter applied to the sandbox.

I did not looked at the code yet, but seccomp should come after landlock.

> For example you run "firejail --landlock firefox" and it should work as before.

Note that reparenting files (move/rename where one of the operates contains a `/`) is always denied by Landlock in ABI version 1.

**Comment by netblue30** [comment]
>>    Run-time detection of kernels 6.1

> Why 6.1? Landlock had nothing new in the 6.1 release.

I'm on Debian stable, default 6.1 kernel. The previous Debian had a 5.9 kernel, but if we find a machine with a lower kernel number and is working fine, we modify the code to accept that version.

>>    Landlock is the last OS filter applied to the sandbox.

>I did not looked at the code yet, but seccomp should come after landlock.

Good point! I'll push it up before seccomp!

>>   For example you run "firejail --landlock firefox" and it should work as before.

>Note that reparenting files (move/rename where one of the operates contains a /) is always denied by Landlock in ABI version 1.

Thanks, I'll look into it.

**Comment by netblue30** [comment]
Note:

> Allow to read and write to files in a directory but not read (list) a directory itself. i.e. ls /etc is forbided, but cat /etc/resolv.conf allowed.

I'll put it here so I don't forget about it. We could probably redo some of the --private features using Landlock, and get rid of the previous implementation. For example: --private-etc, --private-bin, private-opt, and private-srv. I think we have in landlock kernel 6.1 everything we need.

**Reply by rusty-snake** [reply]
Maybe this would even allow use to fix some awkward behaviours of a mount implementation?

> but not read (list) a directory itself. 

Supplement to this, `getdents` is covered by landlock and forbidden. Landlock does not cover `access`/`stat` yet so you can probe the existence of a file with them (`test -e /etc/hosts`).

**Reply by osevan** [reply]
Please dont remove private-etc private-bin and all other private-directory overlay mounts functionality. 

Its better have both - old overlayfs AND new landlocked version .

Because present version is already battle tested and works awesome. 

Better do it as an additional function Like
Landlocked-etc, landlocked-bin, landlocked-dev, landlocked-usb 

And one of most important parts.

landlock "full system path",path2,path3

Separable with , 

And read-write too like landlocked-read-write

And more fine granularity for specific files and folders like read write and execute inside profile. 

Thanks and

Best Regards

**Reply by rusty-snake** [reply]
> Please dont remove

No one said remove. "redo ... implementation"

> overlayfs

There is no overlayfs in the current implementation.

**Comment by kmk3** [comment]
@netblue30 [on Nov 2](https://github.com/netblue30/firejail/discussions/6065#discussioncomment-7458248):

> OK, this is what I have in this moment:
> 
> * Compile time detection based on /usr/include/kernel/landlock.h - if the
>   file is present in the filesystem, the feature is compiled in.
> 
> * Run-time detection of kernels 6.1 (debian stable) or newer - maybe we can
>   push it down to 5.13, but we need to test it.

> I think I'll merge it over the weekend.

Note that the initial commits in the `landlock` branch still contain most of
the main issues raised in #5315, especially the readability-related ones.

Also, even though the branch is based on #5315, there is no attribution given
to @ChrysoliteAzalea in the commits.

Merging the branch directly as is would bring those issues in and make
bisecting harder than it needs to be.

Considering the above, I created a `landlock_v3` branch based on the `landlock`
branch.  I managed to fix many issues, though it's not quite finished.

Diff comparison between the branches:

* <https://github.com/netblue30/firejail/compare/landlock..kmk3:landlock_v3>

Note: The diff only shows the code differences between the entire branches, not
the changes in the commits.

**Comment by netblue30** [comment]
>Considering the above, I created a landlock_v3 branch based on the landlock branch. I managed to fix many issues, though it's not quite finished.

Cool! Let's move on your branch. I don't see it under firejail git. How do we do it? Or we can merge landlock_v3 in the existing landlock branch? Any other ideas?

> Also, even though the branch is based on https://github.com/netblue30/firejail/pull/5315, there is no attribution given to @ChrysoliteAzalea in the commits.

We should fix it and give @ChrysoliteAzalea full attribution.

I still have to put the seccomp fix brought up by @rusty-snake, maybe some other small fixes. And any fixes you guys have, just merge them in directly on the branch, and we'll talk about them on this discussion thread.

**Comment by kmk3** [comment]
@netblue30 [on Nov 6](https://github.com/netblue30/firejail/discussions/6065#discussioncomment-7487306):

> > Considering the above, I created a landlock_v3 branch based on the landlock
> > branch. I managed to fix many issues, though it's not quite finished.
> 
> Cool! Let's move on your branch. I don't see it under firejail git. How do we
> do it? Or we can merge landlock_v3 in the existing landlock branch? Any other
> ideas?

Ah, I had pushed that one to my user.

Pushed the current version to this repository as `landlock_v4`; feel free to
commit on it.

Just let me know when you think it's ready for merging; I'd like to edit a few
details beforehand.

Changes compared to the `landlock` branch:

* <https://github.com/netblue30/firejail/compare/landlock..landlock_v4>

The main change in the code is that it checks whether landlock is supported at
runtime by using a landlock syscall directly, as recommended by landlock(7).

Also, now if `HAVE_LANDLOCK` is not defined, empty `ll_` functions are defined
instead, to allow calling them outside of landlock.c directly (without ifdefs).

> > Also, even though the branch is based on #5315, there is no attribution
> > given to @ChrysoliteAzalea in the commits.
> 
> We should fix it and give @ChrysoliteAzalea full attribution.

Added some details to the first/main landlock commit in the branch.

**Comment by kmk3** [comment]
@netblue30 [on Nov 6](https://github.com/netblue30/firejail/discussions/6065#discussioncomment-7487306):

> I still have to put the seccomp fix brought up by @rusty-snake, maybe some
> other small fixes. And any fixes you guys have, just merge them in directly
> on the branch, and we'll talk about them on this discussion thread.

By the way, if there are going to be multiple people committing in the same
branch, everyone please make sure to configure the following:

```sh
git config pull.ff only
```

Then use `git pull --rebase` when pulling in the branch:

```sh
git checkout landlock_v4
git pull # optional, to check if both local/remote branches have extra commits
git pull --rebase

# If there are conflicts, resolve them then run:
git add -u
git rebase --continue
```

This makes the local `landlock_v4` branch be equal to the remote branch plus
the extra local commits on top.

Which prevents merging the branch into itself when pulling and makes
conflict-related issues on pull easier to spot.  Example:

```
# before
A-B-C-D origin/landlock_v4
   \
    E-F landlock_v4

(git pull --rebase) ->

# after
A-B-C-D origin/landlock_v4
       \
        E-F landlock_v4
```

Also, make sure to not edit (amend/rebase) any commit that was already pushed
to this branch, as that can easily lead to a lot of confusion and conflicts for
whoever else has extra commits on their local branch.

Note: The above is also valid for `master`.

**Comment by netblue30** [comment]
Merged on mainline! seccomp fix and some other small issues are coming.

**Comment by osevan** [comment]
I hope this will be considered

https://github.com/netblue30/firejail/discussions/6065#discussioncomment-7868436

**Comment by osevan** [comment]
Maybe i didnt understand. 
But what was wrong?

Whats the state of landlock inside Firejail? 

Which decisions made?

**Reply by rusty-snake** [reply]
https://github.com/netblue30/firejail/blob/90907ac0d2c34b3c1e9a11f3e5c601308a5b5ff1/src/man/firejail.1.in#L1246-L1295

**Reply by osevan** [reply]
Ok thx.

**Reply by osevan** [reply]
When im understanding right, 

Overlayfs functionality is obsolete with newer kernels and landlock module loading available? 

Or is this only addition ?

**Reply by rusty-snake** [reply]
> Overlayfs functionality is obsolete with newer kernels 

No, but the current implementation is broken on newer kernels. Also it has known security vulnerabilities. That's why it is disabled.

> landlock module loading available

There is no loadable landlock module. See https://www.kernel.org/doc/html/latest/admin-guide/LSM/index.html:

> > The name "module" is a bit of a misnomer since these extensions are not actually loadable kernel modules. Instead, they are selectable at build-time via CONFIG_DEFAULT_SECURITY and can be overridden at boot-time via the "security=..." kernel command line argument, in the case where multiple LSMs were built into a given kernel.

**Reply by rusty-snake** [reply]
Addition to https://github.com/netblue30/firejail/discussions/6065#discussioncomment-7908149,

https://github.com/netblue30/firejail/blob/90907ac0d2c34b3c1e9a11f3e5c601308a5b5ff1/src/man/firejail.1.in#L3418-L3458

**Reply by rusty-snake** [reply]
Second addition, the actual paths for `--landlock` are

https://github.com/netblue30/firejail/blob/90907ac0d2c34b3c1e9a11f3e5c601308a5b5ff1/src/firejail/landlock.c#L210-L237

**Comment by ghost** [comment]
networking support for the Landlock security module landed in kernel 6.7:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=fff69fb03dde

landlock: Support network rules with TCP bind and connect
Add network rules support in the ruleset management helpers and the
landlock_create_ruleset() syscall. Extend user space API to support
network actions:
* Add new network access rights: LANDLOCK_ACCESS_NET_BIND_TCP and
  LANDLOCK_ACCESS_NET_CONNECT_TCP.
* Add a new network rule type: LANDLOCK_RULE_NET_PORT tied to struct
  landlock_net_port_attr. The allowed_access field contains the network
  access rights, and the port field contains the port value according to
  the controlled protocol. This field can take up to a 64-bit value
  but the maximum value depends on the related protocol (e.g. 16-bit
  value for TCP). Network port is in host endianness [1].
* Add a new handled_access_net field to struct landlock_ruleset_attr
  that contains network access rights.

**Reply by rusty-snake** [reply]
#6077

Crablock has already support for it, I will test it when Fedora gets 6.7.

**Reply by rusty-snake** [reply]
Works as expected.

**Comment by amano-kenji** [comment]
What does landlock do? What does it do for firejail?

**Reply by rusty-snake** [reply]
> What does landlock do?

https://www.man7.org/linux/man-pages/man7/landlock.7.html
https://docs.kernel.org/userspace-api/landlock.html

Trying not to go into technical depths:
- Landlock restricts access to the filesystem <sub>(ABIv4 can also restrict TCP Sockets)</sub>.
- It is deny-by-default, you can only access files where access was explicitly granted.
- It works tree based e.g. allow `/home/user/Downloads` means everything under it.
- It is inode (+path_resolution) based rather than path based.
- It is forever, you can not gain access after it got enforce. <sub>File descriptors received from UNIX Sockets work as expected, permissions are checked during `open(2)` not `read(2)`/`write(2)`.</sub>
- Compared to mount-based it has
  - `LANDLOCK_ACCESS_FS_EXECUTE` which maps to `noexec`
  - `LANDLOCK_ACCESS_FS_MAKE_{CHAR,DIR,REG,SOCK,FIFO,BLOCK,SYM}`, `LANDLOCK_ACCESS_FS_REMOVE_{DIR,FILE}`, `LANDLOCK_ACCESS_FS_{WRITE,READ}_FILE` which gives much finer control than `ro`
  - `LANDLOCK_ACCESS_FS_READ_DIR` which does not have a mount equivalent.
- **It still has limitations!**

> [!CAUTION]
> Not all access kinds (syscalls) are covered yet!
> - Metadata is unprotected.
>   - You can not hide the existence of a file. (`access(2)`, `xyz() -> ENOENT`)
>   - You can `stat(2)` every file according to normal permission checks.
>   - ...
> - ...
>
> My full list: https://codeberg.org/crabjail/crablock/src/commit/58262e3218cfcb76c261e25a8ba497823ff9e1d3/man/crablock.1.in.rst#landlock

> What does it do for firejail?

Firejail exposes Landlock ABIv1 based features via `--landlock.*` commands.

**Reply by amano-kenji** [reply]
So, is it a linux security module that comes as a C API that can be used by other LSMs?

**Reply by rusty-snake** [reply]
If the two links above do not help, maybe one of the following links helps:
- https://google.com/
- https://chat.bing.com/

---

> C API

syscalls, nevermind

> linux security module used by other LSMs

LSM == linux security module

It is an own LSM.

Does any LSM use other LSMs (excluding capability)?

**Reply by amano-kenji** [reply]
syscalls are C API from my standpoint... in the same way that suicide is a type of death.

Landlock claims to be a stackable LSM.

**Reply by rusty-snake** [reply]
> stackable LSM

https://lwn.net/Articles/804906/

**Comment by curiosityseeker** [comment]
Thanks, @rusty-snake  - that‘s a very nice overview. 

> Firejail exposes Landlock ABIv1 based features via `--landlock.*` commands.

Yes, but how will this project proceed? Will landlock only be available as an additional optional hardening measure for the time being? Or will all existing profiles be rewritten to integrate landlock rules?

**Reply by rusty-snake** [reply]
> Or will all existing profiles be rewritten to integrate landlock rules?

Adding `include landlock-common.inc` to all profiles we can test:

1. Backward compatibility. It increases the required kernel version by 5 years or so.
2. Forward compatibility. Integration of newer Landlock ABIs.
   - Since Landlock is deny-by-default, adding them to `landlock-common.inc` means to deny more than before.
   - ABIv2: `LANDLOCK_ACCESS_FS_REFER`. Every program that wants to reparent a file (e.g. `rename(2)` where `oldpath` or `newpath` contains a `/` and the program has no `copy_file_range` fallback in case of `EXDEV`) is not supported by ABIv1.
3. Stricter defaults for `landlock-common.inc`?
   - Allow the creation of block devices by default?
   - ...
4. `--landlock.{read,write}` bring them to full power, and allow them on files?

Reimplementing `private*`, `whitelist`, ... with Landlock:

- Landlock is not read yet!
- Even crablock has (still under development) mount based sandboxing to be used in addition to Landlock.
  - Mount: Limit the what
  - Landlock: Limit the how <sub>together with capabilities and seccomp</sub>

**Comment by rusty-snake** [comment]
### More `landlock-common.inc` improvments:

- Stricter `makeipc`
  ```bash
  find / -type p 2>/dev/null
  find / -type s 2>/dev/null | sed -E "s;((/[^/]*){3}).*;\1;" | sort -u
  ```
- `landlock.read /proc` is a useless line
- Missing execute for `/usr/libexec`

**Reply by rusty-snake** [reply]
Minimal example how a .local override for landlock could look like.
```
# d-feet.local

landlock.read /usr
landlock.read /proc
ignore landlock.read
ignore landlock.write
ignore landlock.special
include landlock-common.inc
```

**Comment by curiosityseeker** [comment]
> ignore landlock.read
> ignore landlock.write
> ignore landlock.special
> include landlock-common.inc
> ```

Shouldn't this rather be 
```
ignore landlock.read /
ignore landlock.special /
```
?

**Reply by rusty-snake** [reply]
No.

- `ignore landlock.read /` and `ignore landlock.read` should make no difference (the later will ignore all)
- `landlock.special` (now splitted into `makedev` and `makeipc`) is not required at all. Almost no program needs it.

**Comment by curiosityseeker** [comment]
>     * `landlock.read /proc` is a useless line

Because of 
`landlock.read /`

I guess.

**Reply by rusty-snake** [reply]
Exactly

**Comment by curiosityseeker** [comment]
>     * It is forever, you can not gain access after it got enforce. File descriptors received from UNIX Sockets work as expected, permissions are checked during `open(2)` not `read(2)`/`write(2)`.

Huh? What does this precisely mean? Is this really not reversible? Let's say that I added Landlock rules to a *.local profile and comment them later again - what happens?

**Reply by rusty-snake** [reply]
It means you can not "escape" it via *`fork(2)` or `execve(2)`* because it get inherited and there is no "landlock_unenforce". Just **like `seccomp(2)` or `NO_NEW_PRIVS` work**. Of course nothing that gets enforced by the kernel can live longer than the kernel is running, a reboot will always clear.

**Comment by osevan** [comment]
https://lore.kernel.org/landlock/20240716.yui4Iezai8ae@digikod.net/T/#u

**Comment by osevan** [comment]
https://www.phoronix.com/news/Landlock-Restrictions-Bug-Fixed

Can we resrtrict further fork and keyctl syscalls?

**Reply by rusty-snake** [reply]
- keyctl is already blocked by `seccomp` (`@default`)
- fork (*family!*) can be blocked with `seccomp` to. 

But why do you want?! This vulnerability gets fixed in the Kernel. Do you want to turn off all Kernel features ever used in a vulnerability?

**Reply by osevan** [reply]
Not turn off, but restricting like seccomp.

But ok, when fork can be blocked with seccomp, than seccomp flag inside config file is enough?

Or special flags needed?

**Reply by rusty-snake** [reply]
> Not turn off, but restricting like seccomp.

Turn off in the sandbox.

> But ok, when fork can be blocked with seccomp, than seccomp flag inside config file is enough?

You can use `seccomp` to block syscalls like `seccomp socket` (block all syscalls in the `@default` group and `socket`).

**Reply by osevan** [reply]
Ok thx

**Reply by rusty-snake** [reply]
FTR: Note that `seccomp fork` stops a program from using `fork(2)` (the syscall not the library function) but is not enough to stop a program from forking.

**Reply by rusty-snake** [reply]
FTR: `--rlimit-nproc=` can also be used to limit forking.

**Comment by osevan** [comment]
https://landlock.io/talks/2024-06-06_landlock-article.pdf

"Ongoing work For now, the main limitations of file access control are
metadata access (e.g., file properties, extended attributes) and file path
probing. However, Landlock can fully control access to file contents. We
are working on extending Landlock’s access control with new rights for
already supported subsystems (i.e. file, TCP), but also to support new
subsystems (e.g., task signaling, sockets, IPCs). 12
Performance and usability improvements are also planned, especially
with path walk access caching, tailored data types, and audit support to
improve debuggability and provide metrics [37]"

**Comment by osevan** [comment]
https://www.phoronix.com/news/Landlock-Scoping-Unix-Sockets

**Comment by osevan** [comment]
another question landlock got abi version 6 at the moment, is this already in ?

thank and

best regards

**Comment by osevan** [comment]
i included landlock-common.inc, how i can see if its active ?

**Comment by osevan** [comment]
journalctl -kb -g landlock 
Mar 25 17:29:53 kernel: LSM: initializing lsm=lockdown,capability,landlock,yama,apparmor,tomoyo,bpf,ipe>
Mar 25 17:29:53 kernel: landlock: Up and running.

 firejail --version
firejail version 0.9.73

Compile time support:
        - always force nonewprivs support is disabled
        - AppArmor support is enabled
        - AppImage support is enabled
        - chroot support is enabled
        - D-BUS proxy support is enabled
        - file transfer support is enabled
        - IDS support is disabled
        - Landlock support is enabled
        - networking support is enabled
        - output logging is enabled
        - overlayfs support is disabled
        - private-home support is enabled
        - private-lib support is enabled
        - private-cache and tmpfs as user enabled
        - sandbox check is enabled
        - SELinux support is disabled
        - user namespace support is enabled
        - X11 sandboxing support is enabled



but i want to know if landlock enforced susccesfully
because when i run firejail --debug and my profile it spits out

losing non-standard file descriptors
Starting application
LD_PRELOAD=(null)
Not enforcing Landlock


even it is included the inc file


thanks and

best regards

**Reply by rusty-snake** [reply]
If you do not [enforce](https://man.archlinux.org/man/firejail.1#landlock.enforce) it, it's not enforce.

**Reply by osevan** [reply]
Abi version 6 already inside?

Because we are at the moment on 6.th

**Reply by osevan** [reply]
I have question about landlock.fs.write=path
Some apps needs write data to real fs for persistence.

Can I have access to already wrote data outside of landlocked sandbox?
thanks and

Best regards

**Reply by osevan** [reply]
@rusty-snake can you please answer my question ?

thanks and

best regards

**Reply by rusty-snake** [reply]
> can you please answer my question ?

12 hours, I have a life man. Nevermind.

---

> Abi version 6 already inside?

Only ABI level 1 is supported.

The features of ABI level 4 (`access_net`) and ABI level 6 (`scoped`) are not in the command-line hence they are (oblivious) not used.
The `access_fs` features of ABI level 2 (`REFER`), 3 (`TRUNCATE`) and 5 (`IOCTL_DEV`) aren't supported either (with the expected limitations).

> Can I have access to already wrote data outside of landlocked sandbox?

In order to open a file with `RDONLY` or `RDWR`, the current landlock domain of the calling process must permit `READ_FILE` access on it. So if you only grant `WRITE_FILE`, `open` with `RDWR` will fail, you can only open with `WRONLY`.

**Reply by osevan** [reply]
Ok thanks and I have access of already wrote data inside landlock sandbox when I'm closing landlocked sandbox with exit 0?

Like database for example

**Reply by osevan** [reply]
@rusty-snake, landlock.read already set inside landlock-common.inc to whole fs with / or it is not whole subdirectorys? Need I additionally add landlock.read to subdirectorys too?

**Reply by rusty-snake** [reply]
Landlock uses `path_beneath` rules and are deny-by-default,-then-allow (i.e. there is no `NO_READ_FILE` permission). So a adding an access permission to an directory allows it for everything below this directory. Or the other way around, the access permission for a file are found by walking up its path and adding all permissions that are seen.

**Comment by bar-g** [comment]
> Landlock uses `path_beneath` rules and are deny-by-default,-then-allow 

That is very favorable.

But do I understand it correctly that firejail turns that feature around with its default [landlock-common.inc](https://github.com/netblue30/firejail/blob/master/etc/inc/landlock-common.inc)?

Actually allowing full-system read access, quite critical write access, and more by default?:

```
landlock.fs.read /          # whole system read
landlock.fs.read /proc
landlock.fs.makeipc /       # sockets etc.

# write access
landlock.fs.write ${HOME}
landlock.fs.write ${RUNUSER}
landlock.fs.write /dev
landlock.fs.write /proc
landlock.fs.write /run/shm
landlock.fs.write /tmp

[...]
```

I had hoped for a restrictive "deny all by default and only allow the necessary" mode.

Basicly, the landlock-common.inc being empty and the profiles only adding specific, minimal allow rules. (Possibly loaded from different "app-feature" sub-profiles listing the set of needed permissions required for optional features.)

Would it be be possible for firejail to support that?
Or maybe alternatively, to call a landlock before firejail?

---
PS: It seems the changes originally posted by @rusty-snake in https://github.com/netblue30/firejail/discussions/6065#discussioncomment-8354251 have not been merged into the current landlock-common.inc.

**Reply by kmk3** [reply]
> I had hoped for a restrictive "deny all by default and only allow the
> necessary" mode.

Please see the `LANDLOCK` section in `man firejail` and avoid
landlock-common.inc if you want to apply granular read access.

I'd say that this is currently just an example file for the most part, to make
sure that adding a few landlock commands + `landlock.enforce` work without
breaking the program.

I don't think anyone looked into what would be good default read paths that
work for most programs.

If you can find good specific read paths, feel free to propose them.

Edit: I see now the comment that you mentioned; that could be a good starting
point for improvements.

**Reply by bar-g** [reply]
> I'd say that this is currently just an example file for the most part, to make
sure that adding a few landlock commands + landlock.enforce work without
breaking the program.

Ok, yeah, an example and something to test. So the current version could just be a somehow "bad" example of using landlock, but by all means ok for learning and improve upon.

The current landlock-common.inc seems not to break programs, but from what I read and iterpret, it might too easily collide with the desirable (landlock) "deny by default" policy.

Because when landlock-common.inc ships with `landlock.fs.read /` and users enable it, it seems to allow "everything", overriding all narrower allow rules.
Then depending on `ignore landlock...` rules to mitigate the problems seems suboptimal, because it doesn't err on the safer side.

Instead, I would suggest for discussion what I (naively) thought a common landlock settings include file would do: Only allow very few, most basic, and very specific rules, like `landlock.fs.execute ${path_of_sandboxed_binary}`. And maybe allowances for app-data subdirectories that firejail could create by default, like `~/app/<app-name>`  and `/tmp/app/<app-name>` (to be exported as $TMPDIR environment variable).

Anything else (like specific library file paths etc.) would be the responsibility of the application profiles (to add specific file permissions, and include common rules files for features like a group directory for data exchange).

---

Looking at the current `landlock-common.inc`, maybe it could contain only things that can be considered required "common permissions" _without any doubt_.
The idea is to not let it contain anything that could require `ignore` rules for it to _increase_ sandboxing isolation (something that might be seen as easy to miss, backwards, or just counter-intuitive).
```
# This file is overwritten during software install.
# Persistent customizations should go in a .local file.
include landlock-common.local

##### completely remove these as too broad?
##### (to allow and require limited allowances by profiles)
#landlock.fs.read /          # whole system read
#landlock.fs.read /proc
#landlock.fs.makeipc /       # sockets etc.

# write access
landlock.fs.write ${HOME}    #### restrict this to ${HOME}/app/<app-name>?
landlock.fs.write ${RUNUSER}

#### remove these from common.inc?
#### (let profiles add only specific permissions where really needed)
#landlock.fs.write /dev  
#landlock.fs.write /proc 
#landlock.fs.write /run/shm

landlock.fs.write /tmp      #### restrict this to /tmp/app/<app-name> (put into $TMPDIR)?

# exec access
## misc
landlock.fs.execute ${PATH}    #### restrict to specifically sandboxed binary itself?
# landlock.fs.execute /opt     #### remove from common default

#### remove this from landlock-common.inc, leave to appimage profiles?
#landlock.fs.execute /run/firejail # appimage and various firejail features

#### move this into a landlock-break_wildcard-libs.inc, better let profiles only allow specific files
## lib
#landlock.fs.execute /lib
#landlock.fs.execute /lib32
#landlock.fs.execute /libx32
#landlock.fs.execute /lib64
#landlock.fs.execute /usr/lib
#landlock.fs.execute /usr/lib32
#landlock.fs.execute /usr/libx32
#landlock.fs.execute /usr/lib64
#landlock.fs.execute /usr/local/lib
```

**Comment by curiosityseeker** [comment]
@netblue30 : That Firejail supports landlock is great. However, it still supports only ABI V1 while landlock is now at [ABI V10](https://github.com/landlock-lsm/go-landlock/releases) with great enhancements.

[landrun](https://github.com/zouuup/landrun), e.g., supports a [--best-effort switch](https://github.com/zouuup/landrun#best-effort-mode) - which comes from go-landlock - to ensure compatibility with older kernels. It would be great if something like that would be available in Firejail as well. 

What do you think?

**Reply by osevan** [reply]
This is nice idea.

I hope someone can consider this .

Because on all my sandboxes I'm using 
landlock enforce

Functionality with success .


Thanks and


Best Regards
